Problem Statement
The Chief Operating Officer of a Software Service Provider company is looking to evaluate the efficacy of their incident/issue resolution process and have shared the details of the incident log data. They would like to perform the following analysis to improve their operations to drive better customer satisfaction.

    1. Understand the distribution of incidents to identify the spread by key attributes. 
    2. Understand the alignment between urgency/priority of incidents against the resolution parameters/statistics. 
    3. Build a predictive model using the data that can estimate the resolution time for incident raised in the future.
    4. Build a classification model that would bucket the incidents into high priority/ urgency buckets. 
    5. Suggest recommendations to reduce resolution time.
    
Response: Candidate is expected to respond through a Jupyter notebook or a .py file along with responses on the qualitative questions embedded in the code as comments
Data: Incident Log Data; Data Dictionary 
Evaluation Criteria: The candidate will be evaluated on the following parameters:

    1. Python Knowledge and Quality of code (20%)
    2. Data exploration and processing – with a primary focus on points 1 and 2 above (20%)
    3. Modeling (20%)
    4. Model Evaluation (20%)
    5. Presentation of Results (20%)

# Data column details

1. number: incident identifier (24,918 different values);
2. incident state: eight levels controlling the incident management process transitions from opening until closing the case;
3. active: boolean attribute that shows whether the record is active or closed/canceled;
4. reassignment_count: number of times the incident has the group or the support analysts changed;
5. reopen_count: number of times the incident resolution was rejected by the caller;
6. sys_mod_count: number of incident updates until that moment;
7. made_sla: boolean attribute that shows whether the incident exceeded the target SLA;
8. caller_id: identifier of the user affected;
9. opened_by: identifier of the user who reported the incident;
10. opened_at: incident user opening date and time;
11. sys_created_by: identifier of the user who registered the incident;
12. sys_created_at: incident system creation date and time;
13. sys_updated_by: identifier of the user who updated the incident and generated the current log record;
14. sys_updated_at: incident system update date and time;
15. contact_type: categorical attribute that shows by what means the incident was reported;
16. location: identifier of the location of the place affected;
17. category: first-level description of the affected service;
18. subcategory: second-level description of the affected service (related to the first level description, i.e., to category);
19. u_symptom: description of the user perception about service availability;
20. cmdb_ci: (confirmation item) identifier used to report the affected item (not mandatory);
21. impact: description of the impact caused by the incident (values: 1–High; 2–Medium; 3–Low);
22. urgency: description of the urgency informed by the user for the incident resolution (values: 1–High; 2–Medium; 3–Low);
23. priority: calculated by the system based on 'impact' and 'urgency';
24. assignment_group: identifier of the support group in charge of the incident;
25. assigned_to: identifier of the user in charge of the incident;
26. knowledge: boolean attribute that shows whether a knowledge base document was used to resolve the incident;
27. u_priority_confirmation: boolean attribute that shows whether the priority field has been double-checked;
28. notify: categorical attribute that shows whether notifications were generated for the incident;
29. problem_id: identifier of the problem associated with the incident;
30. rfc: (request for change) identifier of the change request associated with the incident;
31. vendor: identifier of the vendor in charge of the incident;
32. caused_by: identifier of the RFC responsible by the incident;
33. close_code: identifier of the resolution of the incident;
34. resolved_by: identifier of the user who resolved the incident;
35. resolved_at: incident user resolution date and time (dependent variable);
36. closed_at: incident user close date and time (dependent variable).

In [1]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
%matplotlib inline
import matplotlib 

In [2]:
pd.set_option("display.max_columns", 40)

In [3]:
#loading the data set
data = pd.read_csv('incident_response_data/incident_event_log.csv', parse_dates=['opened_at', 'closed_at'])
data.head(50)

number      incident_state  active  reassignment_count  reopen_count  \
0   INC0000045                 New    True                   0             0   
1   INC0000045            Resolved    True                   0             0   
2   INC0000045            Resolved    True                   0             0   
3   INC0000045              Closed   False                   0             0   
4   INC0000047                 New    True                   0             0   
5   INC0000047              Active    True                   1             0   
6   INC0000047              Active    True                   1             0   
7   INC0000047              Active    True                   1             0   
8   INC0000047              Active    True                   1             0   
9   INC0000047              Active    True                   1             0   
10  INC0000047  Awaiting User Info    True                   1             0   
11  INC0000047            Resolved    True                   1             0   
12  INC0000047              Closed   False                   1             0   
13  INC0000057                 New    True                   0             0   
14  INC0000057                 New    True                   0             0   
15  INC0000057                 New    True                   0             0   
16  INC0000057                 New    True                   0             0   
17  INC0000057                 New    True                   0             0   
18  INC0000057            Resolved    True                   0             0   
19  INC0000057              Closed   False                   0             0   
20  INC0000060                 New    True                   0             0   
21  INC0000060              Active    True                   0             0   
22  INC0000060            Resolved    True                   0             0   
23  INC0000060              Closed   False                   0             0   
24  INC0000062                 New    True                   0             0   
25  INC0000062                 New    True                   0             0   
26  INC0000062                 New    True                   1             0   
27  INC0000062                 New    True                   1             0   
28  INC0000062                 New    True                   1             0   
29  INC0000062                 New    True                   1             0   
30  INC0000062            Resolved    True                   1             0   
31  INC0000062              Closed   False                   1             0   
32  INC0000063                 New    True                   0             0   
33  INC0000063                 New    True                   0             0   
34  INC0000063                 New    True                   0             0   
35  INC0000063                 New    True                   0             0   
36  INC0000063                 New    True                   1             0   
37  INC0000063                 New    True                   1             0   
38  INC0000063            Resolved    True                   1             0   
39  INC0000063              Closed   False                   1             0   
40  INC0000064                 New    True                   0             0   
41  INC0000064                 New    True                   0             0   
42  INC0000064              Active    True                   0             0   
43  INC0000064              Active    True                   1             0   
44  INC0000064              Active    True                   1             0   
45  INC0000064              Active    True                   1             0   
46  INC0000064              Active    True                   1             0   
47  INC0000064            Resolved    True                   1             0   
48  INC0000064              Closed   False                   1             0   
49  INC0

In [4]:
len(data.number.unique()) #this is the total no. of incident indentified

20769

# data exploration for selecting the columns for modelling

In [43]:
data.u_priority_confirmation.value_counts()

False    87564
True     32434
Name: u_priority_confirmation, dtype: int64

In [44]:
data.incident_state.value_counts()

Active                33582
New                   30229
Resolved              21500
Closed                20825
Awaiting User Info    12884
Awaiting Vendor         557
Awaiting Problem        400
Awaiting Evidence        19
-100                      2
Name: incident_state, dtype: int64

In [5]:
data.contact_type.value_counts() #it can also be removed as it is skewed

Phone    119879
Email       119
Name: contact_type, dtype: int64

In [6]:
len(data.opened_by.unique())#it is taken as it will provide which user opened the incident

157

In [7]:
len(data.caller_id.unique()) #we can remove caller_id as well as it will cause the curse of dimensionality

4829

In [8]:
data.shape

(119998, 36)

In [9]:
len(data.assignment_group.unique()) #we should keep it to get the performance of the group

73

In [10]:
data.reopen_count.value_counts() #we can remove reopen_count as it is skewed towards 0 and therefore will not much affect our model

0    118044
1      1661
2       141
3        86
4        38
5        11
6         8
7         5
8         4
Name: reopen_count, dtype: int64

In [11]:
data.reassignment_count.value_counts() #it can be removed as it can be adjusted in one hot encoding of assignment group

0     57377
1     32183
2     12961
3      7274
4      4091
5      2310
6      1315
7       900
8       506
9       341
10      272
11      165
12      100
13       57
14       38
15       21
17       16
20       16
16       13
18       13
22        9
19        8
21        3
27        3
23        2
26        2
24        1
25        1
Name: reassignment_count, dtype: int64

In [12]:
data.made_sla.value_counts() #it can also be removed as it is skewed

True     111738
False      8260
Name: made_sla, dtype: int64

In [13]:
len(data.location.unique()) #it is also to be kept to account for the affected location for the incident

203

In [14]:
len(data.category.unique()) #it is also to be kept to account for the affected service

45

In [15]:
data.knowledge.value_counts() #it can also be kept as the data is not skewed and it will account for the knowledge required or not

False    94664
True     25334
Name: knowledge, dtype: int64

# Removing the not so important columns from the data set (columns such as which are giving just the name and time not related to rsolution time or priority and very much skewed columns)

In [16]:
df = data[['number', 'opened_at', 'location', 'impact', 'urgency', 'priority','assignment_group', 'knowledge', 'closed_at']]
df.head(50)

number           opened_at      location      impact     urgency  \
0   INC0000045 2016-02-29 01:16:00  Location 143  2 - Medium  2 - Medium   
1   INC0000045 2016-02-29 01:16:00  Location 143  2 - Medium  2 - Medium   
2   INC0000045 2016-02-29 01:16:00  Location 143  2 - Medium  2 - Medium   
3   INC0000045 2016-02-29 01:16:00  Location 143  2 - Medium  2 - Medium   
4   INC0000047 2016-02-29 04:40:00  Location 165  2 - Medium  2 - Medium   
5   INC0000047 2016-02-29 04:40:00  Location 165  2 - Medium  2 - Medium   
6   INC0000047 2016-02-29 04:40:00  Location 165  2 - Medium  2 - Medium   
7   INC0000047 2016-02-29 04:40:00  Location 165  2 - Medium  2 - Medium   
8   INC0000047 2016-02-29 04:40:00  Location 165  2 - Medium  2 - Medium   
9   INC0000047 2016-02-29 04:40:00  Location 165  2 - Medium  2 - Medium   
10  INC0000047 2016-02-29 04:40:00  Location 165  2 - Medium  2 - Medium   
11  INC0000047 2016-02-29 04:40:00  Location 165  2 - Medium  2 - Medium   
12  INC0000047 2016-02-29 04:40:00  Location 165  2 - Medium  2 - Medium   
13  INC0000057 2016-02-29 06:10:00  Location 204  2 - Medium  2 - Medium   
14  INC0000057 2016-02-29 06:10:00  Location 204  2 - Medium  2 - Medium   
15  INC0000057 2016-02-29 06:10:00  Location 204  2 - Medium  2 - Medium   
16  INC0000057 2016-02-29 06:10:00  Location 204  2 - Medium  2 - Medium   
17  INC0000057 2016-02-29 06:10:00  Location 204  2 - Medium  2 - Medium   
18  INC0000057 2016-02-29 06:10:00  Location 204  2 - Medium  2 - Medium   
19  INC0000057 2016-02-29 06:10:00  Location 204  2 - Medium  2 - Medium   
20  INC0000060 2016-02-29 06:38:00  Location 204  2 - Medium  2 - Medium   
21  INC0000060 2016-02-29 06:38:00  Location 204  2 - Medium  2 - Medium   
22  INC0000060 2016-02-29 06:38:00  Location 204  2 - Medium  2 - Medium   
23  INC0000060 2016-02-29 06:38:00  Location 204  2 - Medium  2 - Medium   
24  INC0000062 2016-02-29 06:58:00   Location 93  2 - Medium  2 - Medium   
25  INC0000062 2016-02-29 06:58:00   Location 93    1 - High  2 - Medium   
26  INC0000062 2016-02-29 06:58:00   Location 93    1 - High  2 - Medium   
27  INC0000062 2016-02-29 06:58:00   Location 93    1 - High  2 - Medium   
28  INC0000062 2016-02-29 06:58:00   Location 93    1 - High  2 - Medium   
29  INC0000062 2016-02-29 06:58:00   Location 93    1 - High  2 - Medium   
30  INC0000062 2016-02-29 06:58:00   Location 93    1 - High  2 - Medium   
31  INC0000062 2016-02-29 06:58:00   Location 93    1 - High  2 - Medium   
32  INC0000063 2016-02-29 07:08:00   Location 93  2 - Medium  2 - Medium   
33  INC0000063 2016-02-29 07:08:00   Location 93  2 - Medium  2 - Medium   
34  INC0000063 2016-02-29 07:08:00   Location 93  2 - Medium  2 - Medium   
35  INC0000063 2016-02-29 07:08:00   Location 93  2 - Medium  2 - Medium   
36  INC0000063 2016-02-29 07:08:00   Location 93  2 - Medium  2 - Medium   
37  INC0000063 2016-02-29 07:08:00   Location 93  2 - Medium  2 - Medium   
38  INC0000063 2016-02-29 07:08:00   Location 93  2 - Medium  2 - Medium   
39  INC0000063 2016-02-29 07:08:00   Location 93  2 - Medium  2 - Medium   
40  INC0000064 2016-02-29 07:10:00  Location 143  2 - Medium  2 - Medium   
41  INC0000064 2016-02-29 07:10:00  Location 143  2 - Medium  2 - Medium   
42  INC0000064 2016-02-29 07:10:00  Location 143  2 - Medium  2 - Medium   
43  INC0000064 2016-02-29 07:10:00  Location 143  2 - Medium  2 - Medium   
44  INC0000064 2016-02-29 07:10:00  Location 143  2 - Medium  2 - Medium   
45  INC0000064 2016-02-29 07:10:00  Location 143  2 - Medium  2 - Medium   
46  INC0000064 2016-02-29 07:10:00  Location 143  2 - Medium  2 - Medium   
47  INC0000064 2016-02-29 07:10:00  Location 143  2 - Medium  2 - Medium   
48  INC0000064 2016-02-29 07:10:00  Location 143  2 - Medium  2 - Medium   
49  INC0000065 2016-02-29 07:38:00  Location 108  2 - Medium  2 - Medium   

        priority assignment_group  knowledge           closed_at  
0   3 - Moderate         Group 56       True 2016-05-03 12:00:00

# Data processing

In [17]:
#finding the incident resolving time
df['resolving_time'] = df.closed_at - df.opened_at
df.head()

/tmp/ipykernel_77862/185017974.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['resolving_time'] = df.closed_at - df.opened_at


number           opened_at      location      impact     urgency  \
0  INC0000045 2016-02-29 01:16:00  Location 143  2 - Medium  2 - Medium   
1  INC0000045 2016-02-29 01:16:00  Location 143  2 - Medium  2 - Medium   
2  INC0000045 2016-02-29 01:16:00  Location 143  2 - Medium  2 - Medium   
3  INC0000045 2016-02-29 01:16:00  Location 143  2 - Medium  2 - Medium   
4  INC0000047 2016-02-29 04:40:00  Location 165  2 - Medium  2 - Medium   

       priority assignment_group  knowledge           closed_at  \
0  3 - Moderate         Group 56       True 2016-05-03 12:00:00   
1  3 - Moderate         Group 56       True 2016-05-03 12:00:00   
2  3 - Moderate         Group 56       True 2016-05-03 12:00:00   
3  3 - Moderate         Group 56       True 2016-05-03 12:00:00   
4  3 - Moderate         Group 70       True 2016-06-03 10:00:00   

    resolving_time  
0 64 days 10:44:00  
1 64 days 10:44:00  
2 64 days 10:44:00  
3 64 days 10:44:00  
4 95 days 05:20:00

In [18]:
df.drop(['opened_at', 'closed_at'], axis=1, inplace=True)
df.head()

/home/shubham/Project/python_ds_env/ds_env/lib/python3.8/site-packages/pandas/core/frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


number      location      impact     urgency      priority  \
0  INC0000045  Location 143  2 - Medium  2 - Medium  3 - Moderate   
1  INC0000045  Location 143  2 - Medium  2 - Medium  3 - Moderate   
2  INC0000045  Location 143  2 - Medium  2 - Medium  3 - Moderate   
3  INC0000045  Location 143  2 - Medium  2 - Medium  3 - Moderate   
4  INC0000047  Location 165  2 - Medium  2 - Medium  3 - Moderate   

  assignment_group  knowledge   resolving_time  
0         Group 56       True 64 days 10:44:00  
1         Group 56       True 64 days 10:44:00  
2         Group 56       True 64 days 10:44:00  
3         Group 56       True 64 days 10:44:00  
4         Group 70       True 95 days 05:20:00

In [19]:
#finding no. of days the incident is active
df.resolving_time = df['resolving_time'].apply(lambda x : int(str(x).split(' ')[0]))
df.head(10)

/home/shubham/Project/python_ds_env/ds_env/lib/python3.8/site-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


number      location      impact     urgency      priority  \
0  INC0000045  Location 143  2 - Medium  2 - Medium  3 - Moderate   
1  INC0000045  Location 143  2 - Medium  2 - Medium  3 - Moderate   
2  INC0000045  Location 143  2 - Medium  2 - Medium  3 - Moderate   
3  INC0000045  Location 143  2 - Medium  2 - Medium  3 - Moderate   
4  INC0000047  Location 165  2 - Medium  2 - Medium  3 - Moderate   
5  INC0000047  Location 165  2 - Medium  2 - Medium  3 - Moderate   
6  INC0000047  Location 165  2 - Medium  2 - Medium  3 - Moderate   
7  INC0000047  Location 165  2 - Medium  2 - Medium  3 - Moderate   
8  INC0000047  Location 165  2 - Medium  2 - Medium  3 - Moderate   
9  INC0000047  Location 165  2 - Medium  2 - Medium  3 - Moderate   

  assignment_group  knowledge  resolving_time  
0         Group 56       True              64  
1         Group 56       True              64  
2         Group 56       True              64  
3         Group 56       True              64  
4         Group 70       True              95  
5         Group 24       True              95  
6         Group 24       True              95  
7         Group 24       True              95  
8         Group 24       True              95  
9         Group 24       True              95

In [20]:
df.priority.value_counts()

3 - Moderate    112115
4 - Low           3549
2 - High          2499
1 - Critical      1835
Name: priority, dtype: int64

Here my <strong>dependent varialbles</strong> are <strong>resolving_time</strong> to which <strong>regression modelling</strong> will be done, and another <strong>dependent variable</strong> is <strong>priority</strong> to which a <strong>classification model</strong> is to be made.

In [21]:
#droping duplicate rows
df.drop_duplicates(inplace=True)
df.head(10)

/home/shubham/Project/python_ds_env/ds_env/lib/python3.8/site-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


number      location      impact     urgency      priority  \
0   INC0000045  Location 143  2 - Medium  2 - Medium  3 - Moderate   
4   INC0000047  Location 165  2 - Medium  2 - Medium  3 - Moderate   
5   INC0000047  Location 165  2 - Medium  2 - Medium  3 - Moderate   
13  INC0000057  Location 204  2 - Medium  2 - Medium  3 - Moderate   
20  INC0000060  Location 204  2 - Medium  2 - Medium  3 - Moderate   
24  INC0000062   Location 93  2 - Medium  2 - Medium  3 - Moderate   
25  INC0000062   Location 93    1 - High  2 - Medium      2 - High   
26  INC0000062   Location 93    1 - High  2 - Medium      2 - High   
32  INC0000063   Location 93  2 - Medium  2 - Medium  3 - Moderate   
36  INC0000063   Location 93  2 - Medium  2 - Medium  3 - Moderate   

   assignment_group  knowledge  resolving_time  
0          Group 56       True              64  
4          Group 70       True              95  
5          Group 24       True              95  
13         Group 70       True              94  
20         Group 25       True             125  
24         Group 70       True              64  
25         Group 70       True              64  
26         Group 23       True              64  
32         Group 70       True              64  
36         Group 23       True              64

In [22]:
df.shape

(36779, 8)

# Now we will do one hot encoding for all categorical features and remove the dependent variables

In [23]:
issue_resolving_time = df[['number','resolving_time']]
issue_resolving_time.drop_duplicates(inplace=True)
issue_resolving_time.head()

/home/shubham/Project/python_ds_env/ds_env/lib/python3.8/site-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


number  resolving_time
0   INC0000045              64
4   INC0000047              95
13  INC0000057              94
20  INC0000060             125
24  INC0000062              64

In [24]:
issue_priority = df[['number', 'priority']]
issue_priority.drop_duplicates(inplace=True)
issue_priority.head()

/home/shubham/Project/python_ds_env/ds_env/lib/python3.8/site-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


number      priority
0   INC0000045  3 - Moderate
4   INC0000047  3 - Moderate
13  INC0000057  3 - Moderate
20  INC0000060  3 - Moderate
24  INC0000062  3 - Moderate

In [25]:
df.drop(['priority', 'resolving_time'], axis=1, inplace=True)
df.head()

/home/shubham/Project/python_ds_env/ds_env/lib/python3.8/site-packages/pandas/core/frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


number      location      impact     urgency assignment_group  \
0   INC0000045  Location 143  2 - Medium  2 - Medium         Group 56   
4   INC0000047  Location 165  2 - Medium  2 - Medium         Group 70   
5   INC0000047  Location 165  2 - Medium  2 - Medium         Group 24   
13  INC0000057  Location 204  2 - Medium  2 - Medium         Group 70   
20  INC0000060  Location 204  2 - Medium  2 - Medium         Group 25   

    knowledge  
0        True  
4        True  
5        True  
13       True  
20       True

now we will do one hot encoding for df

In [26]:
#the function will on hot encode multiple values of open_by, location, category, and assignment_group

def one_hot_encoding(df):
    columns = df.columns
    ini_df = df[columns[0]]
    ini_df.drop_duplicates(inplace=True)
    for i in range(len(columns) - 1):
        sub_df = df[[columns[0], columns[i+1]]]
        sub_df.drop_duplicates(inplace=True)
        sub_ohe_df = pd.crosstab(sub_df[columns[0]],sub_df[columns[i+1]])
        if i == 0:
            ohe_df = pd.merge(ini_df, sub_ohe_df, on='number')
        else:
            ohe_df = pd.merge(ohe_df, sub_ohe_df, on='number')
    return ohe_df


In [27]:
ohe_df = one_hot_encoding(df)
ohe_df.head()

/home/shubham/Project/python_ds_env/ds_env/lib/python3.8/site-packages/pandas/core/series.py:2136: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(result)
/home/shubham/Project/python_ds_env/ds_env/lib/python3.8/site-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


number  ?_x  Location 10  Location 100  Location 101  Location 102  \
0  INC0000045    0            0             0             0             0   
1  INC0000047    0            0             0             0             0   
2  INC0000057    0            0             0             0             0   
3  INC0000060    0            0             0             0             0   
4  INC0000062    0            0             0             0             0   

   Location 107  Location 108  Location 109  Location 11  Location 110  \
0             0             0             0            0             0   
1             0             0             0            0             0   
2             0             0             0            0             0   
3             0             0             0            0             0   
4             0             0             0            0             0   

   Location 111  Location 112  Location 113  Location 114  Location 115  \
0             0             0             0             0             0   
1             0             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             0             0   

   Location 117  Location 118  Location 12  Location 120  ...  Group 66  \
0             0             0            0             0  ...         0   
1             0             0            0             0  ...         0   
2             0             0            0             0  ...         0   
3             0             0            0             0  ...         0   
4             0             0            0             0  ...         0   

   Group 67  Group 68  Group 69  Group 70  Group 71  Group 72  Group 73  \
0         0         0         0         0         0         0         0   
1         0         0         0         1         0         0         0   
2         0         0         0         1         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         0         1         0         0         0   

   Group 74  Group 75  Group 76  Group 77  Group 78  Group 79  Group 80  \
0         0         0         0         0         0         0         0   
1         0         0         0         0         0         0         0   
2         0         0         0         0         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         0         0         0         0         0   

   Group 81  Group 82  Group 9  False  True  
0         0         0        0      0     1  
1         0         0        0      0     1  
2         0         0        0      0     1  
3         0         0        0      0     1  
4         0         0        0      0     1  

[5 rows x 285 columns]

In [28]:
ohe_df.shape

(20769, 285)

# Modelling

# Now lets make a regression model to predict resolving time

In [29]:
X = pd.merge(ohe_df,issue_resolving_time, on='number')
X.head()

number  ?_x  Location 10  Location 100  Location 101  Location 102  \
0  INC0000045    0            0             0             0             0   
1  INC0000047    0            0             0             0             0   
2  INC0000057    0            0             0             0             0   
3  INC0000060    0            0             0             0             0   
4  INC0000062    0            0             0             0             0   

   Location 107  Location 108  Location 109  Location 11  Location 110  \
0             0             0             0            0             0   
1             0             0             0            0             0   
2             0             0             0            0             0   
3             0             0             0            0             0   
4             0             0             0            0             0   

   Location 111  Location 112  Location 113  Location 114  Location 115  \
0             0             0             0             0             0   
1             0             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             0             0   

   Location 117  Location 118  Location 12  Location 120  ...  Group 67  \
0             0             0            0             0  ...         0   
1             0             0            0             0  ...         0   
2             0             0            0             0  ...         0   
3             0             0            0             0  ...         0   
4             0             0            0             0  ...         0   

   Group 68  Group 69  Group 70  Group 71  Group 72  Group 73  Group 74  \
0         0         0         0         0         0         0         0   
1         0         0         1         0         0         0         0   
2         0         0         1         0         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         1         0         0         0         0   

   Group 75  Group 76  Group 77  Group 78  Group 79  Group 80  Group 81  \
0         0         0         0         0         0         0         0   
1         0         0         0         0         0         0         0   
2         0         0         0         0         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         0         0         0         0         0   

   Group 82  Group 9  False  True  resolving_time  
0         0        0      0     1              64  
1         0        0      0     1              95  
2         0        0      0     1              94  
3         0        0      0     1             125  
4         0        0      0     1              64  

[5 rows x 286 columns]

In [30]:
y = X.resolving_time
y.head()

0     64
1     95
2     94
3    125
4     64
Name: resolving_time, dtype: int64

In [31]:
X.drop(['number', 'resolving_time'], axis=1, inplace=True)
X.head()

?_x  Location 10  Location 100  Location 101  Location 102  Location 107  \
0    0            0             0             0             0             0   
1    0            0             0             0             0             0   
2    0            0             0             0             0             0   
3    0            0             0             0             0             0   
4    0            0             0             0             0             0   

   Location 108  Location 109  Location 11  Location 110  Location 111  \
0             0             0            0             0             0   
1             0             0            0             0             0   
2             0             0            0             0             0   
3             0             0            0             0             0   
4             0             0            0             0             0   

   Location 112  Location 113  Location 114  Location 115  Location 117  \
0             0             0             0             0             0   
1             0             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             0             0   

   Location 118  Location 12  Location 120  Location 121  ...  Group 66  \
0             0            0             0             0  ...         0   
1             0            0             0             0  ...         0   
2             0            0             0             0  ...         0   
3             0            0             0             0  ...         0   
4             0            0             0             0  ...         0   

   Group 67  Group 68  Group 69  Group 70  Group 71  Group 72  Group 73  \
0         0         0         0         0         0         0         0   
1         0         0         0         1         0         0         0   
2         0         0         0         1         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         0         1         0         0         0   

   Group 74  Group 75  Group 76  Group 77  Group 78  Group 79  Group 80  \
0         0         0         0         0         0         0         0   
1         0         0         0         0         0         0         0   
2         0         0         0         0         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         0         0         0         0         0   

   Group 81  Group 82  Group 9  False  True  
0         0         0        0      0     1  
1         0         0        0      0     1  
2         0         0        0      0     1  
3         0         0        0      0     1  
4         0         0        0      0     1  

[5 rows x 284 columns]

In [32]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=10)

In [33]:
from sklearn.linear_model import LinearRegression
lr_clf = LinearRegression()
lr_clf.fit(X_train,y_train)
lr_clf.score(X_test,y_test)

/home/shubham/Project/python_ds_env/ds_env/lib/python3.8/site-packages/sklearn/utils/validation.py:1675: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['bool', 'str']. An error will be raised in 1.2.
  warnings.warn(
/home/shubham/Project/python_ds_env/ds_env/lib/python3.8/site-packages/sklearn/utils/validation.py:1675: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['bool', 'str']. An error will be raised in 1.2.
  warnings.warn(


-2.1713042418444416e+20

In [34]:
from sklearn.tree import DecisionTreeRegressor
dt = DecisionTreeRegressor()
dt.fit(X_train,y_train)
dt.score(X_test,y_test)

/home/shubham/Project/python_ds_env/ds_env/lib/python3.8/site-packages/sklearn/utils/validation.py:1675: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['bool', 'str']. An error will be raised in 1.2.
  warnings.warn(
/home/shubham/Project/python_ds_env/ds_env/lib/python3.8/site-packages/sklearn/utils/validation.py:1675: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['bool', 'str']. An error will be raised in 1.2.
  warnings.warn(


-0.14958780422717832

In [35]:
from sklearn.ensemble import RandomForestRegressor
 
 # create regressor object
regressor = RandomForestRegressor(n_estimators = 100, random_state = 0)
 
# fit the regressor with x and y data
regressor.fit(X_train, y_train)
regressor.score(X_test,y_test)

/home/shubham/Project/python_ds_env/ds_env/lib/python3.8/site-packages/sklearn/utils/validation.py:1675: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['bool', 'str']. An error will be raised in 1.2.
  warnings.warn(
/home/shubham/Project/python_ds_env/ds_env/lib/python3.8/site-packages/sklearn/utils/validation.py:1675: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['bool', 'str']. An error will be raised in 1.2.
  warnings.warn(


-0.003010794279248108

# Now make classification model to predict high priority incident

In [36]:
X = pd.merge(ohe_df,issue_priority, on='number')
X.head()

number  ?_x  Location 10  Location 100  Location 101  Location 102  \
0  INC0000045    0            0             0             0             0   
1  INC0000047    0            0             0             0             0   
2  INC0000057    0            0             0             0             0   
3  INC0000060    0            0             0             0             0   
4  INC0000062    0            0             0             0             0   

   Location 107  Location 108  Location 109  Location 11  Location 110  \
0             0             0             0            0             0   
1             0             0             0            0             0   
2             0             0             0            0             0   
3             0             0             0            0             0   
4             0             0             0            0             0   

   Location 111  Location 112  Location 113  Location 114  Location 115  \
0             0             0             0             0             0   
1             0             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             0             0   

   Location 117  Location 118  Location 12  Location 120  ...  Group 67  \
0             0             0            0             0  ...         0   
1             0             0            0             0  ...         0   
2             0             0            0             0  ...         0   
3             0             0            0             0  ...         0   
4             0             0            0             0  ...         0   

   Group 68  Group 69  Group 70  Group 71  Group 72  Group 73  Group 74  \
0         0         0         0         0         0         0         0   
1         0         0         1         0         0         0         0   
2         0         0         1         0         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         1         0         0         0         0   

   Group 75  Group 76  Group 77  Group 78  Group 79  Group 80  Group 81  \
0         0         0         0         0         0         0         0   
1         0         0         0         0         0         0         0   
2         0         0         0         0         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         0         0         0         0         0   

   Group 82  Group 9  False  True      priority  
0         0        0      0     1  3 - Moderate  
1         0        0      0     1  3 - Moderate  
2         0        0      0     1  3 - Moderate  
3         0        0      0     1  3 - Moderate  
4         0        0      0     1  3 - Moderate  

[5 rows x 286 columns]

In [37]:
y = X.priority
y.head()

0    3 - Moderate
1    3 - Moderate
2    3 - Moderate
3    3 - Moderate
4    3 - Moderate
Name: priority, dtype: object

In [38]:
X.drop(['number', 'priority'], axis=1, inplace=True)
X.head()

?_x  Location 10  Location 100  Location 101  Location 102  Location 107  \
0    0            0             0             0             0             0   
1    0            0             0             0             0             0   
2    0            0             0             0             0             0   
3    0            0             0             0             0             0   
4    0            0             0             0             0             0   

   Location 108  Location 109  Location 11  Location 110  Location 111  \
0             0             0            0             0             0   
1             0             0            0             0             0   
2             0             0            0             0             0   
3             0             0            0             0             0   
4             0             0            0             0             0   

   Location 112  Location 113  Location 114  Location 115  Location 117  \
0             0             0             0             0             0   
1             0             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             0             0   

   Location 118  Location 12  Location 120  Location 121  ...  Group 66  \
0             0            0             0             0  ...         0   
1             0            0             0             0  ...         0   
2             0            0             0             0  ...         0   
3             0            0             0             0  ...         0   
4             0            0             0             0  ...         0   

   Group 67  Group 68  Group 69  Group 70  Group 71  Group 72  Group 73  \
0         0         0         0         0         0         0         0   
1         0         0         0         1         0         0         0   
2         0         0         0         1         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         0         1         0         0         0   

   Group 74  Group 75  Group 76  Group 77  Group 78  Group 79  Group 80  \
0         0         0         0         0         0         0         0   
1         0         0         0         0         0         0         0   
2         0         0         0         0         0         0         0   
3         0         0         0         0         0         0         0   
4         0         0         0         0         0         0         0   

   Group 81  Group 82  Group 9  False  True  
0         0         0        0      0     1  
1         0         0        0      0     1  
2         0         0        0      0     1  
3         0         0        0      0     1  
4         0         0        0      0     1  

[5 rows x 284 columns]

In [39]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=10)

In [40]:
from sklearn.ensemble import RandomForestClassifier
randomclassifier=RandomForestClassifier(n_estimators=200,criterion='entropy')
randomclassifier.fit(X_train,y_train)

/home/shubham/Project/python_ds_env/ds_env/lib/python3.8/site-packages/sklearn/utils/validation.py:1675: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['bool', 'str']. An error will be raised in 1.2.
  warnings.warn(


RandomForestClassifier(criterion='entropy', n_estimators=200)

In [41]:
predictions = randomclassifier.predict(X_test)

/home/shubham/Project/python_ds_env/ds_env/lib/python3.8/site-packages/sklearn/utils/validation.py:1675: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['bool', 'str']. An error will be raised in 1.2.
  warnings.warn(


In [42]:
# Import library to check accuracy
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score

matrix=confusion_matrix(y_test,predictions)
print(matrix)
score=accuracy_score(y_test,predictions)
print(score)
report=classification_report(y_test,predictions)
print(report)

[[  31    2   18    3]
 [   5   48   18    0]
 [  31   21 3885    2]
 [   4    1    3  151]]
0.9744257636751125
              precision    recall  f1-score   support

1 - Critical       0.44      0.57      0.50        54
    2 - High       0.67      0.68      0.67        71
3 - Moderate       0.99      0.99      0.99      3939
     4 - Low       0.97      0.95      0.96       159

    accuracy                           0.97      4223
   macro avg       0.77      0.80      0.78      4223
weighted avg       0.98      0.97      0.98      4223



# The classification model to classify the priority of incident is doing great with above 97% accuracy and more than 50% f1-score for all categories
# but the model can if improved to increase the f1 score of the high priority incidents

--------------------------------------------------------------------------------------------------------------------

# The regression model to predict the incident resolution time is not doing much great as it is not giving much accuracy score.
# Some independent variables needs to be selected properly which could account for a higher variation of the dependent variables.
# PCA can also be applied to reduce the dimensions

Sorry not able to increase the accuracy of the regression model due to lack of time

--------------------------------------------------------------------------------------------------------------------

# Some recommendation to reduce resolution time:-
1 Tell the people to double check the priority of an incident and then resolving it. Most people are not doing this according to the data.

2 Don't change the resolver of the assignment group very frequently as the resolver who is solving the incident might know better about the incident with time and solve it quickly. A new resolver may have to some give time in knowing the incident from start and progress which could be saved.

3 Try to minimise the reassignment of the incident from one group to another to resolve the incident quicker.

4 Many a times the resolving time increases due the some outer things like Awaiting User Info, Awaiting Vendor, Awaiting Problem, Awaiting Evidence and not the actual solving of the issue. Try to complete these things before or parallelly with the solving of the incident and don't waste extra time on these things.